k8s/envoy-opa/README.md

```diff
@@ -0,0 +1,357 @@
+
+```console
+import input.attributes.request.http as http_request
+
+valid_path {
+   glob.match("/balances/*", [], http_request.path)
+}
+
+valid_path {
+   glob.match("/profiles/*", [], http_request.path)
+}
+
+valid_path {
+   glob.match("/transactions/*", [], http_request.path)
+}
+```
+
+The function `valid_path` makes use of the built-in function `glob.match(`_pattern_`, `_delimiters_`, `_match_`)` the output of which is true if _match_ can be found in _pattern_ which is separated by _delimiters_. Then, to express logical OR in Rego you define multiple rules with the same name. That is why there are three definitions of `valid_path`, one per valid resource.
+
+The following expression defines that the HTTP method of the request must be equal to `GET`:
+
+```console
+http_request.method == "GET"
+```
+
+And the last expression corresponds also to a user-defined function:
+
+```console
+svc_spiffe_id == "spiffe://example.org/ns/default/sa/default/frontend"
+```
+
+The function `svc_spiffe_id` extracts the SPIFFE ID of the service from the `x-forwarded-client-cert` (XFCC) header in the request. The XFCC header is a proxy header which indicates certificate information for some or all of the clients or proxies that a request has passed through. The `svc_spiffe_id` function leverages two Envoy settings from `envoy.yaml` that modify the HTTP header:
+
+```console
+forward_client_cert_details: sanitize_set
+set_current_client_cert_details:
+   uri: true
+```
+
+When a client connection is mTLS, like this scenario, `forward_client_cert_details: sanitize_set` resets the XFCC header with the client certificate information and `set_current_client_cert_details` specifies the fields in the client certificate to be forwarded.
+
+The XFCC header value is a comma (“,”) separated string. Each substring is an XFCC element and each XFCC element is a semicolon (“;”) separated string. Each substring is a key-value pair, grouped together by an equals (“=”) sign. The following keys are supported by Envoy:
+
+   - `By`        The Subject Alternative Name (URI type) of the current proxy’s certificate.
+   - `Hash`      The SHA 256 digest of the current client certificate.
+   - `Cert`      The entire client certificate in URL encoded PEM format.
+   - `Subject`   The Subject field of the current client certificate. The value is always double-quoted.
+   - `URI`       The URI type Subject Alternative Name field of the current client certificate.
+   - `DNS`       The DNS type Subject Alternative Name field of the current client certificate. A client certificate may contain multiple DNS type Subject Alternative Names, each will be a separate key-value pair.
+
+The following is a sample XFCC header value split into two lines for readability:
+
+```
+By=spiffe://example.org/ns/default/sa/default/backend;Hash=a9317919875e178ce6d61eaa023490
+a2091299753ca5cd01d5323e40696d690b;URI=spiffe://example.org/ns/default/sa/default/frontend
```

Comment:
I'm sorry for iterating on this, but I think putting the literal header name (x-forwarded-client-cert:) helps a little more. I also rebalanced the lines and refrased the introductory sentence.

```suggestion
The following is an XFCC header with a sample value that is split into two lines for readability:

\`\`\`
x-forwarded-client-cert: By=spiffe://example.org/ns/default/sa/default/backend;Hash=a9317919875e178ce6d6
1eaa023490a2091299753ca5cd01d5323e40696d690b;URI=spiffe://example.org/ns/default/sa/default/frontend
\`\`\`
```

Reply:
I do agree with this. It shows a complete example of the header. Looks better. Thanks